File: 卡码/代码随想录/1/02-链表/Java/src/_07_环形链表/LeetCode_0142.java
//142. 环形链表 II
package _07_环形链表;

public class LeetCode_0142 {

}

class ListNode {
	int val;
	ListNode next;

	ListNode() {
	}

	ListNode(int val) {
		this.val = val;
	}

	ListNode(int val, ListNode next) {
		this.val = val;
		this.next = next;
	}
}
class Solution {
  public ListNode detectCycle(ListNode head) {
  	ListNode fast=head,slow=head;
  	while(fast!=null&&fast.next!=null) {
  		fast=fast.next.next;
  		slow=slow.next;
  		//找到相交点则有环
  		if(fast==slow) {
  			ListNode index1=fast,index2=head;
  			while(index1!=index2) {
  				index1=index1.next;
  				index2=index2.next;
  			}
  			return index1;
  		}
  	}
  	return null;
  }
}

File: 卡码/代码随想录/Go/链表/环形链表/LeetCode_0142.go
//142. 环形链表 II
package main

type ListNode struct{
	Val int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	fast,slow:=head,head
	for fast!=nil&&fast.Next!=nil {
		slow=slow.Next
		fast=fast.Next.Next
		//找到相交点则有环
		if slow==fast {
			index1,index2:=fast,head
			for index1!=index2 {
				index1=index1.Next
				index2=index2.Next
			}
			return index1
		}
	}
	return nil
}
